w11b01/test/pgdp/pengusurvivors/ResidualGraphTest.java

```diff
@@ -0,0 +1,178 @@
+        var graphResult = GraphGenerator.generateGraph(
+                verticeString,
+                source,
+                sink
+        );
+        var graph = graphResult.graph();
+        GraphUtils.printGraph(graph);
+        var vertices = graphResult.vertices();
+        graph.generateResidualGraph();
+        var path = graph.findPathInResidual();
+        var matchingPaths = 0;
+        if (expectedPaths.isEmpty()) {
+            if (path != null)
+                throw new AssertionError("Expected no path, but found " + GraphUtils.stringifyPath(path));
+            return;
+        }
+        pathloop:
+        for (var expectedPath : expectedPaths) {
+            if (path == null || expectedPath.size() != path.size()) continue;
+            for (int i = 0; i < path.size(); i++) {
+                var expectedVertex = vertices.get(expectedPath.get(i));
+                var actualVertex = path.get(i);
+                if (expectedVertex != actualVertex) continue pathloop;
+            }
+            matchingPaths++;
+        }
+        assertEquals(1, matchingPaths, "No matching path was found. Your path was: " + (path == null ? "None found" : GraphUtils.stringifyPath(path)));
+
+    }
+
+    static Stream<Arguments> findPathArguments() {
+        return Stream.of(
+                Arguments.of("""
+                        s -> a (1/1);
+                        s -> b (0/1);
+                        a -> c (1/1);
+                        b -> c (0/1);
+                        """, "s", "c", List.of(List.of("s", "b", "c"))),
+                Arguments.of(
+                        """
+                                s -> a;
+                                a -> b;
+                                a -> c;
+                                b -> c;
+                                b -> t;
+                                c -> t;
+                                """,
+                        "s", "b", List.of(List.of("s", "a", "b"))
+                ),
+                Arguments.of(
+                        """
+                                s -> a;
+                                a -> b;
+                                a -> c;
+                                b -> c;
+                                b -> t;
+                                c -> t;
+                                """,
+                        "s", "t", List.of(List.of("s", "a", "b", "t"), List.of("s", "a", "c", "t"), List.of("s", "a", "b", "c", "t"))
+                ),
+                Arguments.of(
+                        """
+                                a -> a;
+                                """
+                        , "a", "a", List.of(List.of("a"))
+                ),
+                Arguments.of(
+                        """
+                                a -> b (1/1);
+                                b -> (0/2);
+                                """,
+                        "a", "c", List.of()
+                )
```

Comment:
Add testcase with potential loop

```suggestion
                ),
                Arguments.of( // tests with potential loops
                        """
                                s -> a;
                                a -> b;
                                a -> t;
                                b -> c;
                                c -> a;
                                b -> t;
                                c -> t;
                                """,
                        "s", "t", List.of(List.of("s", "a", "b", "t"), List.of("s", "a", "t"), List.of("s", "a", "b", "t"), List.of("s", "a", "b", "c", "t"))
                )
```